Add main simulator loop, refactor {Client,Server} and add statistics

irfansharif opened the pull request:
Stripping away the stats handling, the core simulator looks as follows:
```rust
    for i in 0..ticks {
        if client.tick() {
            server.enqueue(Packet(i))
        }
        server.tick()
    }
```
Usage (running qsim -h)
    Usage: qsim [options]

    Options:
        -h, --help          Display this message
            --rate NUM      Average number of generated packets/s (def: 10000)
            --psize NUM     Packet size; bits (def: 1)
            --pspeed NUM    Packet processing speed; bits/s (def: 10000)
            --duration NUM  Duration of simulation; seconds (def: 5)
            --qlimit NUM    Limit on of the buffer queue length; int (def: None)

Sample run:

    Simulation configuration:
      Rate:                  10000 packets/s
      Packet size:           1 bits
      Server speed:          10000 bits/s
      Simulation time:       5s
      Resolution:            1µs
      Queue size limit:      None

    Simulation results:
      Average sojourn time:        0.0168 +/- 0.0084 seconds
      Average # of queued packets: 168.74 +/- 84.45 packets
      Packets generated:           50279 packets
      Packets processed:           49965 packets
      Packets droppped:            0 packets
      Server idle proportion:      0.00%

There are a couple of TODOs around comments and I've not explicitly dealt with the resolution mismatch, though now it should be trivial to add.

irfansharif committed:
Suppress dead code warnings, avoid lossy conversion

irfansharif committed:
Move over Consumer -> simulators::Server

Standardize on the ticker interface where {Client,Server}.tick() moves
the {Client,Server} simulators by one time unit.

Additionally introduce {Server,Client}Statistics and update through
progression of ticks.

Add tests to verify packet generation/delivery/dropping behaviour.

irfansharif committed:
Introduce streaming-stats dependency

Added in order to compute fast averages and variances for packet
sojourn time and buffer utilization.

irfansharif committed:
Add main simulation loop, added more usage flags

Stripping away the stats handling, the core looks as follows:

    for i in 0..ticks {
        if client.tick() {
            server.enqueue(Packet{})
        }
        server.tick()
    }

Usage (running qsim -h)
    Usage: qsim [options]

    Options:
        -h, --help          Display this message
            --rate NUM      Average number of generated packets/s (def: 10000)
            --psize NUM     Packet size; bits (def: 1)
            --pspeed NUM    Packet processing speed; bits/s (def: 10000)
            --duration NUM  Duration of simulation; seconds (def: 5)
            --qlimit NUM    Limit on of the buffer queue length; int (def: None)

Sample run:

    Simulation configuration:
      Rate:                  10000 packets/s
      Packet size:           1 bits
      Server speed:          10000 bits/s
      Simulation time:       5s
      Resolution:            1µs
      Queue size limit:      None

    Simulation results:
      Average sojourn time:        0.0168 +/- 0.0084 seconds
      Average # of queued packets: 168.74 +/- 84.45 packets
      Packets generated:           50279 packets
      Packets processed:           49965 packets
      Packets droppped:            0 packets
      Server idle proportion:      0.00%

irfansharif commented:
Example run from the lab manual: 
```
~ qsim --psize 2000 --pspeed 1000000 --rate 100

Simulation configuration:
         Rate:                  100 packets/s
         Packet size:           2000 bits
         Server speed:          1000000 bits/s
         Simulation time:       5s
         Resolution:            1µs
         Queue size limit:      None

Simulation results:
         Average sojourn time:        0.0013 +/- 0.0008 seconds
         Average # of queued packets: 0.12 +/- 0.36 packets
         Packets generated:           487 packets
         Packets processed:           487 packets
         Packets droppped:            0 packets
         Server idle proportion:      0.81%
```

Dendrimer commented on src/simulators.rs:
```diff
@@ -1,61 +1,192 @@
+        if self.ticker == 0 {
+            self.ticker = self.generator.next_event(self.resolution);
+            let packet = self.queue.pop_front();
+            match packet {
```
Not sure if this is the correct behaviour: if the queue is empty, the ticker shouldn't start since there is technically nothing to "process" at that time... it looks like (correct me if I'm wrong) the ticker starts regardless of whether or not there is actually a packet in the queue.

Dendrimer replied:
In the case that the queue is empty (nothing should be ticking and therefore the server is idle), we still could be stuck in the decrement loop from ticker being restarted by the generator, in which case the idle time is not incrementing and we are not actively for the next packet.

irfansharif replied:
Explained else where, added a comment explaining this pattern:
```rust
// The server operates via polling essentially, we process an entire packet each time the
// ticker returns to zero, if any. If the queue is empty this means the server is idle for that
// time duration and is recorded so internally.
```

Dendrimer commented on src/simulators.rs:
```diff
@@ -1,61 +1,192 @@
+    pub fn tick(&mut self) -> bool {
+        // TODO(irfansharif): Resolution mismatch; no possibility of generating multiple packets.
+        if self.ticker == 0 {
+            self.statistics.packets_generated += 1;
```
Maybe we can express this as 
```rust
self.ticker -= 1
if self.ticker <= 0 { 
   // Code
   true
} else {
   false
}
```
, since this would accomplish the same logic and get rid of this code duplication. It looks like since we are now initializing client to generate a packet at a random time rather than always at 0, the only case where ticker would be 0 at the beginning of a call to ticker would be if the generator ended up generating a 0 due to truncation, in which case we would see a value of -1, which *could* be okay...

Dendrimer replied:
This is more of a nit than anything though

irfansharif replied:
yea the reason was to avoid getting to -1, originally had it that way before getting buffer underflows (ticker is a u32, as I think it should be).

Dendrimer commented:
LGTM

Dendrimer reviewed (approved):
LGTM

jack-mo reviewed (approved):
lgtm

irfansharif committed:
Remove [allow(dead_code)] annotations; added server idle tests

irfansharif committed:
Add comments explaining the {Server,Client}.tick() pattern

The caller is responsible for calling {Client,Server}.tick() at fixed
time intervals, moving the client simulator one time unit per call. For
Client we return a boolean indicating whether or not a packet is
generated in the most recently completed time unit. For Server the
return value, if not None, is the packet that has finished being
processed.
